WatsonWebsocket/WatsonWsClient.cs

```diff
@@ -1,472 +1,813 @@
+            _ClientWs.Options.KeepAliveInterval = TimeSpan.FromSeconds(_KeepAliveIntervalSeconds);
+            if (_PreConfigureOptions != null) _PreConfigureOptions(_ClientWs.Options);
+
+            // Connect
+            _ClientWs.ConnectAsync(_ServerUri, _Token).ContinueWith(AfterConnect);
+        }
+
+        /// <summary>
+        /// Start the client and connect to the server.
+        /// </summary>
+        /// <returns>Task.</returns>
+        public Task StartAsync()
+        {
+            _IsConnecting = true;
+            _Stats = new Statistics();
+            if (_AcceptInvalidCertificates) ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
+            
+            // Pre Configure WebSocket Client
+            _ClientWs.Options.Cookies = _Cookies;
+            _ClientWs.Options.KeepAliveInterval = TimeSpan.FromSeconds(_KeepAliveIntervalSeconds);
+            if (_PreConfigureOptions != null) _PreConfigureOptions(_ClientWs.Options);
+
+            // Connect
+            return _ClientWs.ConnectAsync(_ServerUri, _Token).ContinueWith(AfterConnect);
+        }
+
+        /// <summary>
+        /// Start the client and wait until to connect to the server.
+        /// </summary>
+        /// <param name="token">Cancellation token allowing for termination of this connecting.</param>
+        /// <returns>Boolean indicating if the connection was successfully.</returns>
+        public bool Start(CancellationToken token = default)
+        {
+            _IsConnecting = true;
+            _Stats = new Statistics();
+            if (_AcceptInvalidCertificates) ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
+
+            Stopwatch sw = new Stopwatch();
+            TimeSpan timeOut = TimeSpan.FromSeconds(_ConnectTimeoutSeconds);
+            sw.Start();
+
+            while (sw.Elapsed < timeOut)
+            {
+                if (token.IsCancellationRequested) break;
+                // Reset ClientWebSocket (It dispose it-self)
+                _ClientWs = new ClientWebSocket();
+
+                // Pre Configure WebSocket Client
+                _ClientWs.Options.Cookies = _Cookies;
+                _ClientWs.Options.KeepAliveInterval = TimeSpan.FromSeconds(_KeepAliveIntervalSeconds);
+                if (_PreConfigureOptions != null) _PreConfigureOptions(_ClientWs.Options);
+
+                // Try Connect
+                try
+                {
+                    _ClientWs.ConnectAsync(_ServerUri, token).ContinueWith(AfterConnect).Wait();
+                }
+                catch (System.Net.WebSockets.WebSocketException) { }
+                Thread.Sleep(500); // Wait to Reconnect or Change Connecting State
+
+                // Check Connected
+                if (_ClientWs.State == WebSocketState.Open)
+                {
+                    return true;

```

Comment:
this should also set _IsConnecting back to false if it has connected

Reply:
I removed _IsConnecting (not used)